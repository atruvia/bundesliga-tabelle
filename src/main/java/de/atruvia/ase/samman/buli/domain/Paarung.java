package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.BEENDET;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.GEPLANT;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.LAUFEND;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;

import java.util.function.Function;

import org.jmolecules.ddd.annotation.ValueObject;

import lombok.Builder;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import lombok.With;
import lombok.experimental.Accessors;

@Value
@Builder(toBuilder = true)
@Accessors(fluent = true)
@ValueObject
public class Paarung {

	@Value
	public class PaarungView {

		ViewDirection direction;
		Function<Paarung, Entry> self;
		Function<Paarung, Entry> gegner;
		@Getter(lazy = true)
		Ergebnis ergebnis = calcErgebnis();

		public Entry self() {
			return self.apply(paarung());
		}

		public Entry gegner() {
			return gegner.apply(paarung());
		}

		public int tore() {
			return self().tore();
		}

		public int gegentore() {
			return gegner().tore();
		}

		public boolean isGeplant() {
			return paarung().isGeplant();
		}

		public boolean isLaufend() {
			return paarung().isLaufend();
		}

		private Ergebnis calcErgebnis() {
			var tore = tore();
			var gegentore = gegentore();
			if (tore == gegentore) {
				return UNENTSCHIEDEN;
			} else if (tore > gegentore) {
				return SIEG;
			} else {
				return NIEDERLAGE;
			}
		}

		public ErgebnisTyp ergebnisTyp() {
			return paarung().ergebnisTyp;
		}

		private Paarung paarung() {
			return Paarung.this;
		}

	}

	public enum ViewDirection {
		HEIM, AUSWAERTS
	}

	@RequiredArgsConstructor
	@Getter
	@Accessors(fluent = true)
	public enum Ergebnis {

		SIEG, UNENTSCHIEDEN, NIEDERLAGE;

		public int punkte() {
			return switch (this) {
			case SIEG -> 3;
			case UNENTSCHIEDEN -> 1;
			case NIEDERLAGE -> 0;
			};
		}

	}

	public enum ErgebnisTyp {
		GEPLANT, LAUFEND, BEENDET
	}

	@Value
	@Builder(toBuilder = true)
	public static class Entry {
		Team team;
		@With
		int tore;

		public Entry(Team team, int tore) {
			if (tore < 0) {
				throw new IllegalArgumentException("Tore darf nicht negativ sein, ist aber " + tore);
			}
			this.team = team;
			this.tore = tore;
		}

	}

	@Builder.Default
	ErgebnisTyp ergebnisTyp = GEPLANT;

	Entry heim, gast;

	public boolean isGeplant() {
		return ergebnisTypIs(GEPLANT);
	}

	public boolean isLaufend() {
		return ergebnisTypIs(LAUFEND);
	}

	private boolean ergebnisTypIs(ErgebnisTyp ergebnisTyp) {
		return this.ergebnisTyp == ergebnisTyp;
	}

	public Paarung withErgebnis(int toreHeim, int toreGast) {
		return toBuilder().endergebnis(toreHeim, toreGast).build();
	}

	public ViewDirection viewDirection() {
		return HEIM;
	}

	public static class PaarungBuilder {

		public static PaarungBuilder paarung(String heim, String gast) {
			return paarung(team(heim), team(gast));
		}

		private static Team team(String name) {
			return Team.builder().name(name).build();
		}

		public static PaarungBuilder paarung(Team heim, Team gast) {
			return Paarung.builder().heim(entry(heim)).gast(entry(gast));
		}

		public PaarungBuilder endergebnis(int toreHeim, int toreGast) {
			return ergebnis(BEENDET, toreHeim, toreGast);
		}

		public PaarungBuilder zwischenergebnis(int toreHeim, int toreGast) {
			return ergebnis(LAUFEND, toreHeim, toreGast);
		}

		PaarungBuilder ergebnis(ErgebnisTyp ergebnisTyp, int toreHeim, int toreGast) {
			return ergebnisTyp(ergebnisTyp).goals(toreHeim, toreGast);
		}

		private static Entry entry(Team team) {
			return Entry.builder().team(team).build();
		}

		public PaarungBuilder goals(int toreHeim, int toreGast) {
			return heim(heim.withTore(toreHeim)).gast(gast.withTore(toreGast));
		}

	}

	public PaarungView viewForTeam(ViewDirection viewDirection) {
		return switch (viewDirection) {
		case HEIM -> new PaarungView(HEIM, Paarung::heim, Paarung::gast);
		case AUSWAERTS -> new PaarungView(AUSWAERTS, Paarung::gast, Paarung::heim);
		};
	}

}