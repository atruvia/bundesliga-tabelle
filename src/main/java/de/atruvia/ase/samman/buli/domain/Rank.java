package de.atruvia.ase.samman.buli.domain;

/**
 * Der logische Rank innerhalb der Spielsaison, unabhängig von der Positionierung in der Tabelle
 * (zwei unterschiedliche {@link TabellenPlatz}-Einträge können den gleichen logischen Rank aufweisen).
 */
public record Rank(int value) implements Comparable<Rank> {

    @Override
    public int compareTo(Rank other) {
        return Integer.compare(this.value, other.value);
    }

    @Override
    public String toString() {
        return String.valueOf(value);
    }

}
