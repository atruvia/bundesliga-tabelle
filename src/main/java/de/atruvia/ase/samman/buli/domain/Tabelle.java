package de.atruvia.ase.samman.buli.domain;

import static java.util.Comparator.comparing;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toList;

import java.net.URI;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicInteger;

import de.atruvia.ase.samman.buli.domain.ports.secondary.WappenRepository;
import lombok.RequiredArgsConstructor;
import lombok.ToString;

@RequiredArgsConstructor
public class Tabelle {

	@RequiredArgsConstructor
	@ToString
	private static class OrdnungsElement implements Comparable<OrdnungsElement> {

//	    MIS: Head-to-head goal difference: The goal difference in the matches played between the tied teams.
//	    MIS: Head-to-head away goals scored: The total number of away goals scored by the tied teams in the matches played between them.
//	    HIT: Overall goal difference: The difference between the number of goals scored and the number of goals conceded in all matches played.
//	    HIT: Overall goals scored: The total number of goals scored in all matches played.
//	    HIT: Overall away goals scored: The total number of goals scored by the tied teams in matches played away from their home stadium.
//	    MIS: Overall away goal difference: The difference between the number of away goals scored and the number of away goals conceded by the tied teams in matches played away from their home stadium.
//		If two or more teams have the same rank in the Bundesliga and there is no other criteria that can be used to separate them, then the teams will be listed in alphabetical order according to their full club name.

		private static final Comparator<OrdnungsElement> comparator = comparing(
				(OrdnungsElement e) -> e.tabellenPlatz.getPunkte())
				.thenComparing(comparing(e -> e.tabellenPlatz.getTorDifferenz()))
				.thenComparing(comparing(e -> e.tabellenPlatz.getTore()))
				.thenComparing(comparing(e -> e.tabellenPlatz.getToreAuswaerts())) //
				.reversed();

		private final TabellenPlatz tabellenPlatz;

		@Override
		public int hashCode() {
			return Objects.hash(tabellenPlatz.getPunkte(), tabellenPlatz.getTorDifferenz(), tabellenPlatz.getTore(),
					tabellenPlatz.getToreAuswaerts());
		}

		@Override
		public boolean equals(Object other) {
			return this == other || (other != null && getClass() == other.getClass()
					&& comparator.compare(this, (OrdnungsElement) other) == 0);
		}

		@Override
		public int compareTo(OrdnungsElement other) {
			return comparator.thenComparing(comparing(e -> e.tabellenPlatz.getTeam())).compare(this, other);
		}

	}

	private final Map<String, TabellenPlatz> eintraege = new HashMap<>();
	private final WappenRepository wappenRepository;

	public void add(Paarung paarung) {
		addInternal(paarung, false);
		addInternal(paarung.swap(), true);
	}

	private void addInternal(Paarung paarung, boolean swapped) {
		eintraege.merge(paarung.getTeam1(), newEintrag(paarung, swapped), TabellenPlatz::merge);
	}

	private TabellenPlatz newEintrag(Paarung paarung, boolean swapped) {
		if (!paarung.isGespielt()) {
			return TabellenPlatz.NULL;
		}
		TabellenPlatz.TabellenPlatzBuilder builder = TabellenPlatz.builder() //
				.ergebnis(paarung.ergebnis()) //
				.punkte(paarung.punkte());
		if (swapped) {
			return builder //
					.toreAuswaerts(paarung.getTore()) //
					.gegentoreAuswaerts(paarung.getGegentore()) //
					.build();
		} else
			return builder //
					.toreHeim(paarung.getTore()) //
					.gegentoreHeim(paarung.getGegentore()) //
					.build();
	}

	public List<TabellenPlatz> getEntries() {
		// TODO make it side-affect-free, does it work W/O zip!?
		AtomicInteger platz = new AtomicInteger();
		Map<OrdnungsElement, List<TabellenPlatz>> platzGruppen = eintraege.entrySet().stream().map(this::setTeam)
				.collect(groupingBy(OrdnungsElement::new));
		return platzGruppen.entrySet().stream().sorted(comparing(Entry::getKey)).peek(e -> platz.incrementAndGet())
				.map(Entry::getValue)
				.flatMap(t -> t.stream().sorted(comparing(OrdnungsElement::new)).map(tp -> tp.withPlatz(platz.get()))
						.map(tp -> tp.withWappen(wappen(tp))))
				.collect(toList());
	}

	private URI wappen(TabellenPlatz tp) {
		try {
			return wappenRepository.getWappen("bl1", "2022", tp.getTeam());
		} catch (Exception e) {
			// TODO wenn das Wappen nicht geladen werden kann -> loggen, aber weitermachen
			return null;
		}
	}

	private TabellenPlatz setTeam(Entry<String, TabellenPlatz> entry) {
		return entry.getValue().withTeam(entry.getKey());
	}

}