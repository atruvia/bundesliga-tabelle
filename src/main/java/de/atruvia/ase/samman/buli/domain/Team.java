package de.atruvia.ase.samman.buli.domain;

public class Team {
	
	String

}
