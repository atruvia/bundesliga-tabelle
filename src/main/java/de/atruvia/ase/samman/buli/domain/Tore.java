package de.atruvia.ase.samman.buli.domain;

import org.jmolecules.ddd.annotation.ValueObject;

import lombok.Value;
import lombok.experimental.Accessors;

@Value
@Accessors(fluent = true)
@ValueObject
public class Tore implements Comparable<Tore> {

	public static final Tore NULL = new Tore(0);

	int anzahl;

	public static Tore tore(int anzahl) {
		if (anzahl < 0) {
			throw new IllegalArgumentException("Value must not be negative bus was " + anzahl);
		}
		return new Tore(anzahl);
	}

	public Tore add(Tore other) {
		return new Tore(anzahl + other.anzahl);
	}

	public Tore minus(Tore other) {
		return new Tore(anzahl - other.anzahl);
	}

	@Override
	public int compareTo(Tore other) {
		return Integer.compare(anzahl, other.anzahl);
	}

}
