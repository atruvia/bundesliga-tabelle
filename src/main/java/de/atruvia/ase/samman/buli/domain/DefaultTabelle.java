package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;
import static java.util.Collections.emptyList;
import static java.util.Comparator.comparing;
import static java.util.function.Function.identity;
import static java.util.stream.Collectors.groupingBy;
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Stream.concat;
import static lombok.AccessLevel.PRIVATE;

import java.util.Collection;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.TreeMap;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

import org.jmolecules.ddd.annotation.ValueObject;

import de.atruvia.ase.samman.buli.domain.Paarung.PaarungView;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import lombok.experimental.Accessors;

@RequiredArgsConstructor
@ValueObject
public class DefaultTabelle implements Tabelle {

	@Value
	@Accessors(fluent = true)
	private static class OrdnungsElement implements Comparable<OrdnungsElement> {

		private static final Comparator<OrdnungsElement> comparator = comparing(value(TabellenPlatz::punkte)) //
				.thenComparing(value(TabellenPlatz::torDifferenz)) //
				.thenComparing(value(TabellenPlatz::gesamtTore)) //
				.thenComparing(direkterVergleichGesamt()) //
				.thenComparing(direkterVergleichAuswaertsTore()) //
				.thenComparing(value(TabellenPlatz::auswaertsTore)) //
				.reversed();

		private static Comparator<OrdnungsElement> direkterVergleichGesamt() {
			return compareWithSwapped((that, other) -> whereToreIs(that.tabellenPlatz, gegnerIs(other.team())));
		}

		private static Comparator<OrdnungsElement> direkterVergleichAuswaertsTore() {
			return compareWithSwapped(
					(that, other) -> whereToreIs(that.tabellenPlatz, gegnerIs(other.team()).and(isAuswaerts())));
		}

		private static <T, R extends Comparable<R>> Comparator<T> compareWithSwapped(BiFunction<T, T, R> biFunction) {
			return (o1, o2) -> biFunction.apply(o1, o2).compareTo(biFunction.apply(o2, o1));
		}

		private Team team() {
			return tabellenPlatz.team();
		}

		private static Predicate<PaarungView> gegnerIs(Team gegner) {
			return e -> Objects.equals(e.gegner().team().id(), gegner.id());
		}

		private static Predicate<PaarungView> isAuswaerts() {
			return e -> Objects.equals(e.direction(), AUSWAERTS);
		}

		private static int whereToreIs(TabellenPlatz tabellenPlatz, Predicate<PaarungView> filter) {
			return tabellenPlatz.paarungenStream().filter(filter).map(PaarungView::tore).mapToInt(Integer::valueOf)
					.sum();
		}

		@Getter(value = PRIVATE)
		TabellenPlatz tabellenPlatz;

		@Override
		public int hashCode() {
			return 0;
		}

		private static <T> Function<OrdnungsElement, T> value(Function<TabellenPlatz, T> function) {
			return t -> function.apply(t.tabellenPlatz);
		}

		@Override
		public boolean equals(Object o) {
			return comparator.compare(this, (OrdnungsElement) o) == 0;
		}

		@Override
		public int compareTo(OrdnungsElement other) {
			return comparator.thenComparing(e -> e.team().name()).compare(this, other);
		}

	}

	private final Collection<TabellenPlatz> entries;

	public DefaultTabelle() {
		this(emptyList());
	}

	@Override
	public Tabelle add(Paarung paarung) {
		var existing = entries().stream();
		var toAdd = Stream.of(newEntry(paarung.viewForTeam(HEIM)), newEntry(paarung.viewForTeam(AUSWAERTS)));
		var newEntries = concat(existing, toAdd) //
				.collect(toMap(t -> t.team().id(), identity(), TabellenPlatz::mergeWith)) //
				.values();
		return new DefaultTabelle(newEntries);
	}

	private static TabellenPlatz newEntry(PaarungView paarung) {
		var builder = TabellenPlatz.builder().team(paarung.self().team());
		return (paarung.isGeplant() ? builder : builder.paarung(paarung)).build();
	}

	public List<TabellenPlatz> entries() {
		// TODO make it side-effect-free
		AtomicInteger counter = new AtomicInteger(1);
		Map<OrdnungsElement, List<TabellenPlatz>> platzGruppen = entries.stream()
				.collect(groupingBy(OrdnungsElement::new));
		return new TreeMap<>(platzGruppen).values().stream() //
				.flatMap(t -> makeGroup(counter, t)) //
				.toList();
	}

	private static Stream<TabellenPlatz> makeGroup(AtomicInteger counter, List<TabellenPlatz> tabellenPlaetze) {
		Rank rank = new Rank(counter.getAndAdd(tabellenPlaetze.size()));
		return tabellenPlaetze.stream() //
				.sorted(comparing(OrdnungsElement::new)) //
				.map(tp -> tp.withRank(rank));
	}

}
