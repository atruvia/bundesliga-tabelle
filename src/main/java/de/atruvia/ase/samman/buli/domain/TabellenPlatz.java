package de.atruvia.ase.samman.buli.domain;

import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp.BEENDET;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.AUSWAERTS;
import static de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection.HEIM;
import static de.atruvia.ase.samman.buli.util.Merger.lastNonNull;
import static de.atruvia.ase.samman.buli.util.Merger.merge;
import static java.util.Arrays.asList;
import static java.util.stream.IntStream.range;

import java.net.URI;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BinaryOperator;
import java.util.stream.Stream;

import de.atruvia.ase.samman.buli.domain.Paarung.Ergebnis;
import de.atruvia.ase.samman.buli.domain.Paarung.ErgebnisTyp;
import de.atruvia.ase.samman.buli.domain.Paarung.PaarungView;
import de.atruvia.ase.samman.buli.domain.Paarung.ViewDirection;
import de.atruvia.ase.samman.buli.util.Merger.Mergeable;
import lombok.Builder;
import lombok.Value;
import lombok.With;
import lombok.experimental.Accessors;

@Value
@Accessors(fluent = true)
@Builder
public class TabellenPlatz implements Mergeable<TabellenPlatz> {

	@Value
	private static class ErgebnisEntry {
		Ergebnis ergebnis;
		ErgebnisTyp ergebnisTyp;
	}

	private static final BinaryOperator<Integer> adder = (i1, i2) -> i1 + i2;

	URI wappen;
	@With
	int platz;
	@With
	String team;
	int spiele;
	List<ErgebnisEntry> ergebnisse;
	int punkte;
	Map<ViewDirection, Integer> tore;
	Map<ViewDirection, Integer> gegentore;
	PaarungView laufendesSpiel;

	public List<Ergebnis> ergebnisse() {
		return collectToList(ergebnisseStream());
	}

	public List<Ergebnis> ergebnisse(ErgebnisTyp... ergebnisTyp) {
		return collectToList(ergebnisseStream().filter(e -> entryErgebnisIsTypeOf(e, ergebnisTyp)));
	}

	private Stream<ErgebnisEntry> ergebnisseStream() {
		return ergebnisse.stream();
	}

	private static List<Ergebnis> collectToList(Stream<ErgebnisEntry> filter) {
		return filter.map(ErgebnisEntry::ergebnis).toList();
	}

	private static boolean entryErgebnisIsTypeOf(ErgebnisEntry e, ErgebnisTyp... ergebnisTyp) {
		return asList(ergebnisTyp).contains(e.ergebnisTyp());
	}

	public int gesamtTore() {
		return heimtore() + auswaertsTore();
	}

	public int gesamtGegentore() {
		return heimGegentore() + auswaertsGegentore();
	}

	public int heimtore() {
		return tore().getOrDefault(HEIM, 0);
	}

	public int auswaertsTore() {
		return tore().getOrDefault(AUSWAERTS, 0);
	}

	public int heimGegentore() {
		return gegentore().getOrDefault(HEIM, 0);
	}

	public int auswaertsGegentore() {
		return gegentore().getOrDefault(AUSWAERTS, 0);
	}

	public int torDifferenz() {
		return gesamtTore() - gesamtGegentore();
	}

	public TabellenPlatzBuilder toBuilder() {
		TabellenPlatzBuilder builder = new TabellenPlatzBuilder();
		builder.wappen = wappen;
		builder.platz = platz;
		builder.team = team;
		builder.spiele = spiele;
		builder.ergebnisse = new ArrayList<>(ergebnisse);
		builder.punkte = punkte;
		builder.tore = new HashMap<>(tore);
		builder.gegentore = new HashMap<>(gegentore);
		builder.laufendesSpiel = laufendesSpiel;
		return builder;
	}

	public static class TabellenPlatzBuilder {

		public TabellenPlatzBuilder() {
			ergebnisse = new ArrayList<>();
			tore = new HashMap<>();
			gegentore = new HashMap<>();
		}

		public TabellenPlatzBuilder ergebnis(Ergebnis ergebnis, ErgebnisTyp ergebnisTyp) {
			ergebnisse.add(new ErgebnisEntry(ergebnis, ergebnisTyp));
			return this;
		}

		public TabellenPlatzBuilder withGegentore(ViewDirection direction, int anzahl) {
			gegentore.put(direction, anzahl);
			return this;
		}

		public TabellenPlatzBuilder withTore(ViewDirection viewDirection, int anzahl) {
			tore.put(viewDirection, anzahl);
			return this;
		}

	}

	@Override
	public TabellenPlatz mergeWith(TabellenPlatz other) {
		return builder() //
				.team(lastNonNull(team, other.team)) //
				.ergebnisse(merge(ergebnisse, other.ergebnisse)) //
				.spiele(merge(spiele, other.spiele)) //
				.punkte(merge(punkte, other.punkte)) //
				.tore(merge(adder, tore, other.tore)) //
				.gegentore(merge(adder, gegentore, other.gegentore)) //
				.wappen(lastNonNull(wappen, other.wappen)) //
				.laufendesSpiel(lastNonNull(laufendesSpiel, other.laufendesSpiel)) //
				.build();
	}

	public int siege() {
		return countAnzahl(SIEG);
	}

	public int unentschieden() {
		return countAnzahl(UNENTSCHIEDEN);
	}

	public int niederlagen() {
		return countAnzahl(NIEDERLAGE);
	}

	private int countAnzahl(Ergebnis type) {
		return (int) ergebnisseStream().map(ErgebnisEntry::ergebnis).filter(type::equals).count();
	}

	public Ergebnis[] tendenz() {
		return last(5, ergebnisse(BEENDET)).toArray(Ergebnis[]::new);
	}

	private static <T> Stream<T> last(int count, List<T> elements) {
		int size = elements.size();
		return range(0, count).mapToObj(i -> {
			int index = size - i - 1;
			return index >= 0 ? elements.get(index) : null;
		});
	}

}