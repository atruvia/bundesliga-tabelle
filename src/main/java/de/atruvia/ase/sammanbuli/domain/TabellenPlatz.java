package de.atruvia.ase.sammanbuli.domain;

import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.NIEDERLAGE;
import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.SIEG;
import static de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis.UNENTSCHIEDEN;
import static java.util.Collections.emptyMap;
import static java.util.stream.Collectors.toMap;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Map.Entry;
import java.util.stream.Stream;

import de.atruvia.ase.sammanbuli.domain.Paarung.Ergebnis;
import lombok.Builder;
import lombok.Value;
import lombok.With;

@Value
@Builder
public class TabellenPlatz {

	static TabellenPlatz NULL = new TabellenPlatz(0, "", 0, emptyMap(), 0, 0, 0, 0, 0);

	@With
	int platz;
	@With
	String team;
	@Builder.Default
	int spiele = 1;
	Map<Ergebnis, Integer> ergebnisse;
	int punkte;
	int toreHeim;
	int toreAuswaerts;
	int gegentoreHeim;
	int gegentoreAuswaerts;

	public int getTore() {
		return toreHeim + toreAuswaerts;
	}

	public int getGegentore() {
		return gegentoreHeim + gegentoreAuswaerts;
	}

	static class TabellenPlatzBuilder {

		TabellenPlatzBuilder() {
			ergebnisse = new HashMap<>();
		}

		public TabellenPlatz.TabellenPlatzBuilder ergebnis(Ergebnis ergebnis) {
			ergebnisse.merge(ergebnis, 1, (a, b) -> a + b);
			return this;
		}

	}

	public int getTorDifferenz() {
		return getTore() - getGegentore();
	}

	public TabellenPlatz merge(TabellenPlatz other) {
		return TabellenPlatz.builder() //
				.ergebnisse(merge(this.ergebnisse, other.ergebnisse)) //
				.spiele(this.spiele + other.spiele) //
				.punkte(this.punkte + other.punkte) //
				.toreHeim(this.toreHeim + other.toreHeim) //
				.toreAuswaerts(this.toreAuswaerts + other.toreAuswaerts) //
				.gegentoreHeim(this.gegentoreHeim + other.gegentoreHeim) //
				.gegentoreAuswaerts(this.gegentoreAuswaerts + other.gegentoreAuswaerts) //
				.build();
	}

	private static Map<Ergebnis, Integer> merge(Map<Ergebnis, Integer> map1, Map<Ergebnis, Integer> map2) {
		return Stream.of(map1, map2).map(Map::entrySet).flatMap(Collection::stream)
				.collect(toMap(Entry::getKey, Entry::getValue, Integer::sum));
	}

	public int getGewonnen() {
		return ergebnis(SIEG);
	}

	public int getUnentschieden() {
		return ergebnis(UNENTSCHIEDEN);
	}

	public int getVerloren() {
		return ergebnis(NIEDERLAGE);
	}

	private int ergebnis(Ergebnis type) {
		return ergebnisse.getOrDefault(type, 0);
	}

}