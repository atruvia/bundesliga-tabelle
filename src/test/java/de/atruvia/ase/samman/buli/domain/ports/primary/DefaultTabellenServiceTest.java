package de.atruvia.ase.samman.buli.domain.ports.primary;

import static de.atruvia.ase.samman.buli.infra.adapters.secondary.OpenLigaDbSpieltagRepoMother.*;
import static java.util.stream.Collectors.joining;
import static org.approvaltests.Approvals.verify;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;

import org.junit.jupiter.api.Test;

import de.atruvia.ase.samman.buli.domain.TabellenPlatz;

class DefaultTabellenServiceTest {

	@Test
	void tabelleBl12022Spieltag24() {
		TabellenService sut = new DefaultTabellenService(spieltagFsRepo(), wappenFsRepo());
		List<TabellenPlatz> erstelleTabelle = sut.erstelleTabelle("bl1", "2022");
		String tabelle = erstelleTabelle.stream().map(f -> print(f, longestTeamName(erstelleTabelle)))
				.collect(joining("\n"));
		verify(tabelle);
	}

	private int longestTeamName(List<TabellenPlatz> tabellenPlaetze) {
		return tabellenPlaetze.stream().map(TabellenPlatz::getTeam).mapToInt(String::length).max().orElse(0);
	}

	private String print(TabellenPlatz tabellenPlatz, int length) {
		return Arrays.asList(stringFormat(length, tabellenPlatz.getTeam()), tabellenPlatz.getSpiele(),
				tabellenPlatz.getSiege(), tabellenPlatz.getUnentschieden(), tabellenPlatz.getNiederlagen(),
				tabellenPlatz.getTore(), tabellenPlatz.getGegentore(), tabellenPlatz.getTorDifferenz(),
				tabellenPlatz.getPunkte()).stream().map(this::format).collect(joining(" | "));
	}

	private String stringFormat(int length, String team) {
		return String.format("%-" + (length + 1) + "s", team);
	}

	private String format(Object o) {
		return o instanceof Number ? String.format("%3d", o) : Objects.toString(o);
	}

}
