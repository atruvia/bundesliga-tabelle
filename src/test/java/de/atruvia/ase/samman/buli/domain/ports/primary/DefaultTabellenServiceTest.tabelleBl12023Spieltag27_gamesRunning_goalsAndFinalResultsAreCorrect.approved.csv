mannschaft,spiele,siege,unentschieden,niederlagen,tore,gegentore,torDifferenz,punkte,tendenz,laufendesSpiel,wappen
Bayer Leverkusen,27,23,4,0,68,19,49,73,SSSSS,,https://upload.wikimedia.org/wikipedia/de/thumb/f/f7/Bayer_Leverkusen_Logo.svg/1200px-Bayer_Leverkusen_Logo.svg.png
FC Bayern München,27,19,3,5,78,33,45,60,NSSUS,,https://i.imgur.com/jJEsJrj.png
VfB Stuttgart,27,18,3,6,63,34,29,57,USSSU,,https://i.imgur.com/v0tkpNx.png
Borussia Dortmund,27,15,8,4,55,32,23,53,SSSSN,,https://upload.wikimedia.org/wikipedia/commons/thumb/6/67/Borussia_Dortmund_logo.svg/560px-Borussia_Dortmund_logo.svg.png
RB Leipzig,27,15,5,7,60,32,28,50,USSSN,,https://i.imgur.com/Rpwsjz1.png
Eintracht Frankfurt,28,10,12,6,43,36,7,42,UNSSU,1:1(Werder Bremen),https://i.imgur.com/X8NFkOb.png
FC Augsburg,27,9,9,9,44,43,1,36,USSSS,,https://i.imgur.com/sdE62e2.png
SC Freiburg,27,10,6,11,39,48,-9,36,SNSUN,,https://i.imgur.com/r3mvi0h.png
TSG 1899 Hoffenheim,27,9,6,12,45,52,-7,33,NNNSS,,https://i.imgur.com/gF0PfEl.png
Werder Bremen,28,8,7,13,36,44,-8,31,NNNNU,1:1(Eintracht Frankfurt),https://upload.wikimedia.org/wikipedia/commons/thumb/b/be/SV-Werder-Bremen-Logo.svg/681px-SV-Werder-Bremen-Logo.svg.png
1. FC Heidenheim 1846,27,7,9,11,38,47,-9,30,UUNNU,,https://upload.wikimedia.org/wikipedia/commons/thumb/9/9d/1._FC_Heidenheim_1846.svg/830px-1._FC_Heidenheim_1846.svg.png
1. FC Union Berlin,27,8,5,14,25,42,-17,29,USNNU,,https://assets.dfb.de/uploads/000/018/232/small_union-Berlin.jpg
Borussia Mönchengladbach,27,6,10,11,46,53,-7,28,NUUUS,,https://i.imgur.com/KSIk0Eu.png
VfL Wolfsburg,27,7,7,13,33,44,-11,28,SNNNU,,https://i.imgur.com/ucqKV4B.png
VfL Bochum,27,5,11,11,32,56,-24,26,UNNNN,,https://i.imgur.com/5jy3Gfr.png
1. FSV Mainz 05,27,3,11,13,22,46,-24,20,USNUN,,https://upload.wikimedia.org/wikipedia/commons/thumb/9/9e/Logo_Mainz_05.svg/1200px-Logo_Mainz_05.svg.png
1. FC Köln,27,3,10,14,21,48,-27,19,UNUNU,,https://upload.wikimedia.org/wikipedia/commons/thumb/0/01/1._FC_Koeln_Logo_2014–.svg/296px-1._FC_Koeln_Logo_2014–.svg.png
SV Darmstadt 98,27,2,8,17,28,67,-39,14,UNNNU,,https://upload.wikimedia.org/wikipedia/commons/e/e5/SV_Darmstadt_98_Logo.svg